### DISCUSSION TITLE
Difference between /api and /pages/api

### DISCUSSION THREAD

**Discussion by fsa317** [vercel/community#1998] [Help] [answered]
### Summary

Per this [previous discussion](https://github.com/orgs/vercel/discussions/1698), I am having issues with using both python and node serverless functions inside the /pages/api directory but it works inside the /api directory.  This had been working fine for me but I recently tried to integrate with next auth and its giving me an error "Request not defined" which appears to be related to my code being in /api and not /pages/api

So I have two questions:
1. What is the difference between api and pages/api?
2. Which is the proper setup to have both python and node functions?

**Comment by amyegan** [comment, marked as answer]
The difference between `/api` and `/pages/api` is ownership, in a sense. Vercel [Serverless Function Runtimes](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes#) handle Node, Python, Go, and Ruby functions in the `/api` directory.  While Next.js handles only Node functions in `/pages/api`.  Because of these differences in routing ownership, conflicts can happen and lead to unexpected behavior. 

If you can easily convert the Python functions into JS, then they can all be placed in `/page/api` for your Next.js site. If you can't easily convert them, or simply don't want to, you can deploy the Python functions to their own project. Then you can [use rewrites](https://vercel.com/guides/can-i-redirect-from-a-subdomain-to-a-subpath) or [add CORS headers](https://vercel.com/guides/how-to-enable-cors) to get all of the pieces connected.

**Reply by fsa317** [reply]
Ok, thanks for the clarification, seems like I need to do some refactoring.  It's odd because I have it working right now with everything in `/api` but that seems to cause issues when using other libraries that expect things in `/pages/api`
